[Style] Add a shared base class for Style WebCore::Length wrappers to reduce code duplication

weinig opened the pull request:
#### 7832e782324e1abbf200594249833acf672cea5c
<pre>
[Style] Add a shared base class for Style WebCore::Length wrappers to reduce code duplication
<a href="https://bugs.webkit.org/show_bug.cgi?id=294811">https://bugs.webkit.org/show_bug.cgi?id=294811</a>

Reviewed by Darin Adler.

Adds a shared base class for strong Style types that wrap WebCore::Length
that is customizable via template parameters allowing for a nice reduction
in code duplication.

Eventually, these should all use Style::PrimitiveNumericOrKeyword&lt;&gt;, but
to ensure we have no performance regressions, we are first moving to
&quot;zero-cost&quot; wrappers of the existing type, WebCore::Length, to enforce
the type safety without any memory layout changes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Length.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
* Source/WebCore/rendering/RenderTextInlines.h:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/style/values/box/StyleMargin.cpp:
* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/box/StylePadding.cpp:
* Source/WebCore/style/values/box/StylePadding.h:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.cpp:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.h:
* Source/WebCore/style/values/motion/StyleOffsetDistance.cpp:
* Source/WebCore/style/values/motion/StyleOffsetDistance.h:
* Source/WebCore/style/values/position/StyleInset.cpp:
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/sizing/StyleMaximumSize.cpp:
* Source/WebCore/style/values/sizing/StyleMaximumSize.h:
* Source/WebCore/style/values/sizing/StyleMinimumSize.cpp:
* Source/WebCore/style/values/sizing/StyleMinimumSize.h:
* Source/WebCore/style/values/sizing/StylePreferredSize.cpp:
* Source/WebCore/style/values/sizing/StylePreferredSize.h:

Canonical link: <a href="https://commits.webkit.org/296527@main">https://commits.webkit.org/296527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c408bdd5a0509b31df1bec2970bf9e2831fb5ce8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108797 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114007 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59166 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110760 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82670 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23158 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63109 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16134 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58709 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117129 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26476 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91682 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36223 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91492 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23295 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36388 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14144 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31727 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35749 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41283 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35455 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38800 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
```diff
@@ -0,0 +1,287 @@
+    LengthWrapperBase(Fixed&& fixed) : m_value(fixed.value, WebCore::LengthType::Fixed) { }
+    LengthWrapperBase(const Fixed& fixed) : m_value(fixed.value, WebCore::LengthType::Fixed) { }
+    LengthWrapperBase(Percentage&& percent) : m_value(percent.value, WebCore::LengthType::Percent) { }
+    LengthWrapperBase(const Percentage& percent) : m_value(percent.value, WebCore::LengthType::Percent) { }
```
Given none of these do an expensive copy, do we need the rvalue-reference overloads? Or if it is an expensive copy then should the rvalue-reference ones be doing a move?

weinig replied:
No, you are right. No need for the two overloads. Can just be a single value parameter one.

darinadler commented on Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
```diff
@@ -0,0 +1,287 @@
+    auto operator()(const T& value, float referenceLength) -> float
+    {
+        return floatValueForLength(value.m_value, referenceLength);
+    }
```
```suggestion
    auto operator()(const T& value, LayoutUnit referenceLength) -> LayoutUnit
    {
        return valueForLength(value.m_value, referenceLength);
    }
    auto operator()(const T& value, float referenceLength) -> float
    {
        return floatValueForLength(value.m_value, referenceLength);
    }
```

darinadler reviewed (approved):
Fantastic to get rid of >1000 unneeded lines of code!

weinig committed:
[Style] Add a shared base class for Style WebCore::Length wrappers to reduce code duplication
https://bugs.webkit.org/show_bug.cgi?id=294811

Reviewed by Darin Adler.

Adds a shared base class for strong Style types that wrap WebCore::Length
that is customizable via template parameters allowing for a nice reduction
in code duplication.

Eventually, these should all use Style::PrimitiveNumericOrKeyword<>, but
to ensure we have no performance regressions, we are first moving to
"zero-cost" wrappers of the existing type, WebCore::Length, to enforce
the type safety without any memory layout changes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/Length.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
* Source/WebCore/rendering/RenderTextInlines.h:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/style/values/box/StyleMargin.cpp:
* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/box/StylePadding.cpp:
* Source/WebCore/style/values/box/StylePadding.h:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.cpp:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.h:
* Source/WebCore/style/values/motion/StyleOffsetDistance.cpp:
* Source/WebCore/style/values/motion/StyleOffsetDistance.h:
* Source/WebCore/style/values/position/StyleInset.cpp:
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/sizing/StyleMaximumSize.cpp:
* Source/WebCore/style/values/sizing/StyleMaximumSize.h:
* Source/WebCore/style/values/sizing/StyleMinimumSize.cpp:
* Source/WebCore/style/values/sizing/StyleMinimumSize.h:
* Source/WebCore/style/values/sizing/StylePreferredSize.cpp:
* Source/WebCore/style/values/sizing/StylePreferredSize.h:

Canonical link: https://commits.webkit.org/296527@main
